Rust/src/visitor/visitor.rs

```diff
@@ -0,0 +1,38 @@
+use crate::{Internal, Leaf};
+
+/// Visitors on random cut trees.
+///
+/// Many random cut forest scoring algorithms use a visitor design pattern on
+/// the nodes of a tree. The nodes typically run from the leaf closest to the
+/// input query point along the path to the root node such as in the function
+/// [`Tree::traverse()`](crate::tree::Tree::traverse).
+///
+pub trait Visitor<T> {
```

Comment:
Do we need to define both `T` and `Output` in the trait definition?

Reply:
I believe so. `T` is (almost) always the float type. You can see it show up in the definition of this trait's `accept()` function:

```rust
fn accept(&mut self, node: &Internal<T>, depth: T);
```

I use a separate associated output type to handle the attribution visitor case. In many situations the compiler can infer `T`. For example, in `RandomCutForest::anomaly_score()` you see the construction line,

```rust
let mut visitor = AnomalyScoreVisitor::new(sampled_tree.tree(), point);
```

Since `sampled_tree` is of type `SampledTree<T>` the `T` part is inferred. That is, I don't need to write `let mut visitor: AnomalyScoreVisitor<T>`. The type of the output is defined in the visitor implementation. For example, in the definition of `AnomalyScoreVisitor` we have,

```rust
impl<'a, T> Visitor<T> for AnomalyScoreVisitor<'a, T> where
    T: Float + One + Sum + Zero
{
    type Output = T;
    // ...
}
```

When I finally create an `AttributionScoreVisitor` I would set `type Output = Vec<T>`. This output type can also be inferred in some cases since `Self::Output` is the output type of `get_result()`.

Reply:
One follow up question, though, is whether `depth` should be of type `T`. I do so for convenience since depth is used to update the scalar anomaly score which is of type `T`. But it could be of type `u16`, `u32`, or whatever. I'm leaning towards leaving it at `T` but would like to hear your opinion.